firedrake/plot.py

```diff
@@ -1,373 +1,280 @@
 
+
+def trisurf(function, *args, **kwargs):
+    r"""Create a 3D surface plot of a finite element field
+
+    :arg function: the function to plot
+    :return:
+    """
+    axes = kwargs.pop("axes", None)
     if axes is None:
-        axes = plt.subplot(111)
+        figure = plt.figure()
+        axes = figure.add_subplot(projection='3d')
 
-    def display_plot(index):
-        axes.clear()
-        return plot(functions[index], num_sample_points, axes=axes, **kwargs).figure
+    if len(function.ufl_shape) == 1:
+        mesh = function.ufl_domain()
+        element = function.ufl_element().sub_elements()[0]
+        Q = FunctionSpace(mesh, element)
+        function = interpolate(sqrt(inner(function, function)), Q)
 
-    interact(display_plot, index=IntSlider(min=0, max=len(functions)-1,
-                                           step=1, value=0),
-             continuous_update=False)
-    return axes
+    num_sample_points = kwargs.pop("num_sample_points", 10)
+    triangulation, vals = _two_dimension_triangle_func_val(function,
+                                                           num_sample_points)
 
+    _kwargs = {"antialiased": False, "edgecolor": "none", "shade": False,
+               "cmap": plt.rcParams["image.cmap"]}
+    _kwargs.update(kwargs)
 
-def piecewise_linear(points, axes=None, **kwargs):
-    """Plot a piece-wise linear plot for the given points, returns a
-    matplotlib axes.
+    return axes.plot_trisurf(triangulation, vals, *args, **_kwargs)
 
-    :arg points: Points to be plotted
-    :arg axes: Axes to be plotted on
-    :arg kwargs: Additional key word arguments passed to plot
+
+def tripcolor(function, *args, **kwargs):
+    r"""Create a pseudo-color plot of a finite element field
+
+    :arg function: the function to plot
+    :return: matplotlib Collection object
+    """
+    return _plot_2d_field("tripcolor", function, *args, **kwargs)
+
+
+def quiver(function, *args, **kwargs):
+    r"""Make a quiver plot of a vector field
+
+    :arg function: the vector field to plot
+    :return: matplotlib Quiver object
     """
-    try:
-        import matplotlib.pyplot as plt
-    except ImportError:
-        raise RuntimeError("Matplotlib not importable, is it installed?")
+    if function.ufl_shape != (2,):
+        raise ValueError("Quiver plots only defined for 2D vector fields!")
+
+    axes = kwargs.pop("axes", None)
+    if axes is None:
+        figure = plt.figure()
+        axes = figure.add_subplot()
+
+    coords = function.ufl_domain().coordinates.dat.data_ro
+    V = function.ufl_domain().coordinates.function_space()
+    vals = interpolate(function, V).dat.data_ro
+    C = np.linalg.norm(vals, axis=1)
+    return axes.quiver(*(coords.T), *(vals.T), C, *args, **kwargs)
+
+
+def plot(function, *args, bezier=False, num_sample_points=10, **kwargs):
+    r"""Plot a Firedrake object.
+
+    :arg function_or_mesh: The :class:`~.Function` to plot
```

Comment:
This can be updated now to note limitations. Presumably "Firedrake object" is now too broad a term.

There's also a bunch of undocumented arguments here that should be documented